# Exercise 4.8 Follow Along

### Importing Libraries and Setup

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# creating path variable
path = r'/Users/paigepearson/Desktop/Data Analytics/Immersion Course/Achievement 4/Instacart Basket Analysis'

In [3]:
# importing orders_products_merged_updated.pkl dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [4]:
# creating data subset
df = ords_prods_merge[:1000000]

In [5]:
# checking the subset for irregularities
df.shape

(1000000, 19)

In [6]:
df.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

  price_range_loc     busiest_day  busiest_days busiest_period_of_day  
0       Mid-range  Regularly busy  Regular days           Most orders  
1       Mid-range  Regularly busy  Slowest days        Average orders  
2       Mid-range  Regularly busy  Slowest days           Most orders  
3       Mid-range      Least busy  Slowest days        Average orders  
4       Mid-range      Least busy  Slowest days           Most orders

Observation: my data subset has 19 columns, the lesson example shows 14 columns. Do I get rid of the busiest_day? that wasn't explicitly stated in 4.7, will consult tutor.

### Grouping & Aggregating Data

In [7]:
# grouping by product name, once again the output is slightly different than the lesson example, figure 3
df.groupby('product_name')

In [8]:
# grouping by department id, setup step 1 and 2 out of 3 
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# the mean() function can be used to produce the same results
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
# performing multiple aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [11]:
# creating a loyalty flag column, step 1 - group based on 'user_id', step 2 apply the transform() function on 'order_number' column to generate max orders per user, step 3 create new column 'max_order' and put results in this column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN         196                  1   
1                   7                   15.0         196                  1   
2                  12                   21.0         196                  1   
3                   7                   29.0         196                  1   
4                  15                   28.0         196                  1   
5                   7                   19.0         196                  1   
6                   9                   20.0         196                  1   
7                  14                   14.0         196                  2   
8                  16                    0.0         196                  4   
9                   8                   30.0         196                  1   
10                  9                    7.0         196                  2   
11                 16                    8.0         196                  6   
12                 11                    7.0         196                  1   
13                  9                    7.0         196                  1   
14                 10                   14.0         196                  1   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
5           1         Soda        77              7     9.0   both   
6           1         Soda        77              7     9.0   both   
7           1         Soda        77              7     9.0   both   
8           1         Soda        77              7     9.0   both   
9           1         Soda        77              7     9.0   both   
10          0         Soda        77              7     9.0   both   
11          1         Soda        77              7     9.0   both   
12          1         Soda        77              7     9.0   both   
13          1         Soda        77              7     9.0   both   
14          1         Soda        77              7     9.0   both   

   price_range_loc     busiest_day  busiest_days busiest_period_of_day  \
0        Mid-range  Regularly busy  Regular days           Most orders   
1        Mid-range  Regularly busy  Slowest days        Average orders   
2        Mid-range  Regularly busy  Slowest days           Most orders   
3        Mid-range      Least busy  Slowest days        Average orders   
4        Mid-range      Least busy  Slowest days           Most orders   
5        Mid-range  Regularly busy  Regular days        Average orders   
6        Mid-range  Regular

In [13]:
# setting the flags
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
# checking work
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [17]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

In [18]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... aisle_id  department_id  prices  _merge price_range_loc  \
0          0  ...       77              7     9.0    both       Mid-range   
1          1  ...       77              7     9.0    both       Mid-range   
2          1  ...       77              7     9.0    both       Mid-range   
3          1  ...       77              7     9.0    both       Mid-range   
4          1  ...       77              7     9.0    both       Mid-range   

      busiest_day  busiest_days busiest_period_of_day max_order  loyalty_flag  
0  Regularly busy  Regular days           Most orders        10  New customer  
1  Regularly busy  Slowest days        Average orders        10  New customer  
2  Regularly busy  Slowest days           Most orders        10  New customer  
3      Least busy  Slowest days        Average orders        10  New customer  
4      Least busy  Slowest days           Most orders        10  New customer  

[5 rows x 21 columns]

# Task 4.8

## Step 1 - Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

Note: See above for step 1

## Step 2 - In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [19]:
# finding the aggregated mean of 'order_number' grouped by department_id
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Observations: As the subset was smaller, all 21 departments were not featured in the results. Likewise, the number of entries made the mean results vary. In the results from the entire dataframe, all of the mean values dropped except for the household department (department_id 17), which went from 11.2 to 15.6. 

## Step 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

Note: Please see Grouping & Aggregating Data for step 4 results

## Step 5 - Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [20]:
# checking statistics of loyalty_flag for information on product prices
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Observations: The min and max don't offer much as it shows that at least one customer as bought both the lowest and highest priced products. The mean shows that the more purchases a customer has the trend is for their mean to go down, probably just balancing out the lower amount purchases with the higher amount ones. New customers don't have nearly as many purchases so their mean is most likely more varied from customer to customer. It strikes me as odd that the max is a perfect 99999.0 when the mean is so much lower. It might be an error of some kind skewing the data or just one large ticket item.

In [21]:
# sorting value in 'prices' column to see what the top prices look like.
ords_prods_merge['prices'].sort_values(ascending = False)

29165979    99999.0
29165985    99999.0
29165519    99999.0
29165518    99999.0
29165517    99999.0
             ...   
7870751         1.0
7870750         1.0
7870749         1.0
7870748         1.0
30034999        1.0
Name: prices, Length: 32404859, dtype: float64

In [22]:
# trying to get a sense of prices
ords_prods_merge['prices'].value_counts(dropna = False)

12.3    624261
10.3    510496
4.0     447272
6.3     438467
1.3     425691
         ...  
20.2       123
22.7        99
21.5        85
18.3        11
21.0         8
Name: prices, Length: 242, dtype: int64

In [23]:
# creating a subset to see if I can spot the higher prices
df_price_hunt = ords_prods_merge.loc[ords_prods_merge['prices'] > 50]

In [24]:
df_price_hunt.head(20)

order_id  user_id eval_set  order_number  order_day_of_week  \
10030345    912404       17    prior            12                  2   
10030346    603376       17    prior            22                  6   
10030347   3264360      135    prior             2                  2   
10030348    892534      135    prior             3                  0   
10030349    229704      342    prior             8                  1   
10030350   2856927      618    prior             2                  5   
10030351   1871776      618    prior             3                  6   
10030352   2575782      658    prior            19                  1   
10030353    642211      658    prior            21                  3   
10030354    591140      658    prior            30                  6   
10030355   1116834      658    prior            32                  2   
10030356    260601      846    prior             2                  3   
10030357    573149      866    prior             4                  3   
10030358   2589688      866    prior             5                  3   
10030359    525880      866    prior             6                  1   
10030360    265986      866    prior             7                  1   
10030361   2920240      866    prior             8                  2   
10030362    500920      866    prior             9                  0   
10030363   1457337      866    prior            10                  5   
10030364   1916506      866    prior            11                  1   

          order_hour_of_day  days_since_last_order  product_id  \
10030345                 14                    5.0       21553   
10030346                 16                    4.0       21553   
10030347                 21                   13.0       21553   
10030348                  8                   12.0       21553   
10030349                 19                   30.0       21553   
10030350                 12                   30.0       21553   
10030351                 13                   15.0       21553   
10030352                 15                    9.0       21553   
10030353                 17                   21.0       21553   
10030354                 10                   12.0       21553   
10030355                 21                   13.0       21553   
10030356                 11                    3.0       21553   
10030357                 14                   23.0       21553   
10030358                 12                   21.0       21553   
10030359                 11                   12.0       21553   
10030360                  9                    7.0       21553   
10030361                 10                   15.0       21553   
10030362                 12                   19.0       21553   
10030363                 14                   12.0       21553   
10030364                 15                   17.0       21553   

          add_to_cart_order  reordered  ... aisle_id  department_id   prices  \
10030345                  5          0  ...      108             16  14900.0   
10030346                  3          1  ...      108             16  14900.0   
10030347                  6          0  ...      108             16  14900.0   
10030348                  3          1  ...      108             16  14900.0   
10030349                  9          0  ...      108             16  14900.0   
10030350                  5          0  ...      108             16  14900.0   
10030351                  9          1  ...      108             16  14900.0   
10030352                 21          0  ...      108             16  14900.0   
10030353                 17          1  ...      108             16  14900.0   
10030354                 17          1  ...      108             16  14900.0   
10030355                 17          1  ...      108             16  14900.0   
10030356                  5          0  ...      108             16  14900.0   
10030357                 22          0  ...      

In [25]:
df_price_hunt.describe()

order_id        user_id  order_number  order_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000        5127.000000   
mean   1.708876e+06  106117.185684     15.302126           2.750536   
std    9.838956e+05   58594.982659     16.691849           2.010892   
min    4.300000e+01      17.000000      1.000000           0.000000   
25%    8.642320e+05   58136.000000      4.000000           1.000000   
50%    1.693799e+06  106021.000000      9.000000           3.000000   
75%    2.567244e+06  156389.000000     20.000000           5.000000   
max    3.421047e+06  206049.000000     99.000000           6.000000   

       order_hour_of_day  days_since_last_order    product_id  \
count        5127.000000            4776.000000   5127.000000   
mean           13.473376              11.899079  23201.815682   
std             4.137326               9.068019   4153.742747   
min             0.000000               0.000000  21553.000000   
25%            10.000000               5.000000  21553.000000   
50%            13.000000               8.000000  21553.000000   
75%            16.000000              16.000000  21553.000000   
max            23.000000              30.000000  33664.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        5127.000000  5127.000000  5127.000000         5127.0   
mean            7.986347     0.613614   104.732592           16.0   
std             6.868889     0.486968     8.231346            0.0   
min             1.000000     0.000000    84.000000           16.0   
25%             3.000000     0.000000   108.000000           16.0   
50%             6.000000     1.000000   108.000000           16.0   
75%            11.000000     1.000000   108.000000           16.0   
max            72.000000     1.000000   108.000000           16.0   

             prices    max_order  
count   5127.000000  5127.000000  
mean   26485.547494    29.791691  
std    29186.636448    24.929207  
min    14900.000000     3.000000  
25%    14900.000000    12.000000  
50%    14900.000000    21.000000  
75%    14900.000000    41.000000  
max    99999.000000    99.000000

Observations: Now I've found two prices that are on the vastly higher end, product_id 21553 with the price of 14,900 dollars, and product_id 33664 with a price of 99,999 dollars. Both are in the dairy/eggs department. The current dataframe does not include product names, and I need to import the original products.csv to find the final part of the information.

In [26]:
# importing original products.csv
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [27]:
# locating the two product id's I found earlier
df_prods[df_prods['product_id'] == 21553]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

        prices  
21554  14900.0

In [28]:
df_prods[df_prods['product_id'] == 33664]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

Observation: I'm pretty sure Instacart doesn't sell the most expensive milk and cottage cheese in the nation, and that these two prices are an error. Will ask for advice from tutor as to how to proceed. If I could I would send out for a price check from Instacart to change the price to the correct amount. However, seeing as I may not have this information for this project I may opt to create another csv with these two items removed to help balance the data. For now I won't touch it for the scope of this project.

## Step 6 - Create a spending flag for each user based on the average price across all their orders

In [29]:
# grouping average order prices by user_id and entering data into a newly created column 'average_spent'
ords_prods_merge['average_spent'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [30]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... department_id  prices  _merge  price_range_loc  \
0          0  ...             7     9.0    both        Mid-range   
1          1  ...             7     9.0    both        Mid-range   
2          1  ...             7     9.0    both        Mid-range   
3          1  ...             7     9.0    both        Mid-range   
4          1  ...             7     9.0    both        Mid-range   

      busiest_day  busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy  Regular days           Most orders        10  New customer   
1  Regularly busy  Slowest days        Average orders        10  New customer   
2  Regularly busy  Slowest days           Most orders        10  New customer   
3      Least busy  Slowest days        Average orders        10  New customer   
4      Least busy  Slowest days           Most orders        10  New customer   

   average_spent  
0       6.367797  
1       6.367797  
2       6.367797  
3       6.367797  
4       6.367797  

[5 rows x 22 columns]

In [31]:
# creating spending flag
ords_prods_merge.loc[ords_prods_merge['average_spent'] < 10, 'spending_flag'] = 'Low spender'

In [32]:
ords_prods_merge.loc[ords_prods_merge['average_spent'] >= 10, 'spending_flag'] = 'High spender'

In [33]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... prices  _merge  price_range_loc     busiest_day  \
0          0  ...    9.0    both        Mid-range  Regularly busy   
1          1  ...    9.0    both        Mid-range  Regularly busy   
2          1  ...    9.0    both        Mid-range  Regularly busy   
3          1  ...    9.0    both        Mid-range      Least busy   
4          1  ...    9.0    both        Mid-range      Least busy   

   busiest_days busiest_period_of_day max_order  loyalty_flag average_spent  \
0  Regular days           Most orders        10  New customer      6.367797   
1  Slowest days        Average orders        10  New customer      6.367797   
2  Slowest days           Most orders        10  New customer      6.367797   
3  Slowest days        Average orders        10  New customer      6.367797   
4  Slowest days           Most orders        10  New customer      6.367797   

   spending_flag  
0    Low spender  
1    Low spender  
2    Low spender  
3    Low spender  
4    Low spender  

[5 rows x 23 columns]

In [34]:
# checking work, spending flag counts
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

## Step 7 - Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [35]:
# grouping median days_since_last_order by user_id and creating a column for the results
ords_prods_merge['median_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [36]:
# checking work
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... _merge  price_range_loc     busiest_day  busiest_days  \
0          0  ...   both        Mid-range  Regularly busy  Regular days   
1          1  ...   both        Mid-range  Regularly busy  Slowest days   
2          1  ...   both        Mid-range  Regularly busy  Slowest days   
3          1  ...   both        Mid-range      Least busy  Slowest days   
4          1  ...   both        Mid-range      Least busy  Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_spent spending_flag  \
0           Most orders        10  New customer      6.367797   Low spender   
1        Average orders        10  New customer      6.367797   Low spender   
2           Most orders        10  New customer      6.367797   Low spender   
3        Average orders        10  New customer      6.367797   Low spender   
4           Most orders        10  New customer      6.367797   Low spender   

   median_orders  
0           20.5  
1           20.5  
2           20.5  
3           20.5  
4           20.5  

[5 rows x 24 columns]

In [37]:
# creating order_frequency_flag
ords_prods_merge.loc[ords_prods_merge['median_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [38]:
ords_prods_merge.loc[(ords_prods_merge['median_orders'] > 10) & (ords_prods_merge['median_orders'] <=20), 'order_frequency_flag'] = 'Regular customer'

In [39]:
ords_prods_merge.loc[ords_prods_merge['median_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [40]:
# checking work
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... price_range_loc     busiest_day  busiest_days  \
0          0  ...       Mid-range  Regularly busy  Regular days   
1          1  ...       Mid-range  Regularly busy  Slowest days   
2          1  ...       Mid-range  Regularly busy  Slowest days   
3          1  ...       Mid-range      Least busy  Slowest days   
4          1  ...       Mid-range      Least busy  Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_spent spending_flag  \
0            Most orders        10  New customer      6.367797   Low spender   
1         Average orders        10  New customer      6.367797   Low spender   
2            Most orders        10  New customer      6.367797   Low spender   
3         Average orders        10  New customer      6.367797   Low spender   
4            Most orders        10  New customer      6.367797   Low spender   

  median_orders   order_frequency_flag  
0          20.5  Non-frequent customer  
1          20.5  Non-frequent customer  
2          20.5  Non-frequent customer  
3          20.5  Non-frequent customer  
4          20.5  Non-frequent customer  

[5 rows x 25 columns]

In [41]:
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

Note: Unsure how to access the NaNs to investigate further, still puttering with this to figure out how to view the 5  NaN values. Need to view NaNs that show up in a single column vs. the entire df, as the NaNs from days_since_last_order are clogging up my results.

In [44]:
ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()]

order_id  user_id eval_set  order_number  order_day_of_week  \
13645692    895835   159838    prior             1                  0   
17251990    895835   159838    prior             1                  0   
17622767    895835   159838    prior             1                  0   
24138593    895835   159838    prior             1                  0   
25880002    895835   159838    prior             1                  0   

          order_hour_of_day  days_since_last_order  product_id  \
13645692                 17                    NaN       10749   
17251990                 17                    NaN       33401   
17622767                 17                    NaN       23695   
24138593                 17                    NaN       21334   
25880002                 17                    NaN       22198   

          add_to_cart_order  reordered  ... price_range_loc  busiest_day  \
13645692                  3          0  ...       Mid-range  Busiest day   
17251990                  6          0  ...       Mid-range  Busiest day   
17622767                  2          0  ...       Low-range  Busiest day   
24138593                  5          0  ...       Mid-range  Busiest day   
25880002                  1          0  ...       Low-range  Busiest day   

          busiest_days  busiest_period_of_day max_order  loyalty_flag  \
13645692  Busiest days            Most orders         1  New customer   
17251990  Busiest days            Most orders         1  New customer   
17622767  Busiest days            Most orders         1  New customer   
24138593  Busiest days            Most orders         1  New customer   
25880002  Busiest days            Most orders         1  New customer   

         average_spent spending_flag median_orders  order_frequency_flag  
13645692          7.42   Low spender           NaN                   NaN  
17251990          7.42   Low spender           NaN                   NaN  
17622767          7.42   Low spender           NaN                   NaN  
24138593          7.42   Low spender           NaN                   NaN  
25880002          7.42   Low spender           NaN                   NaN  

[5 rows x 25 columns]

Observation: the 5 NaNs also coincide with median_orders having NaN values, the user_id is the same, this was their first purchase on the site. This might be an issue with this user specifically, or possibly any user who has only made a single purchase. 

In [48]:
ords_prods_merge[ords_prods_merge.isnull().any(axis=1)]

order_id  user_id eval_set  order_number  order_day_of_week  \
0          2539329        1    prior             1                  2   
23         2497897       52    prior             1                  1   
37         3205960       67    prior             1                  3   
58         1889835       82    prior             1                  3   
75          520620      120    prior             1                  3   
...            ...      ...      ...           ...                ...   
32404835   2608655   197349    prior             1                  3   
32404840    216145   205587    prior             1                  3   
32404843    609055   202557    prior             1                  0   
32404856    758936   203436    prior             1                  2   
32404858   3093936   205420    prior             1                  4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
23                        9                    NaN         196   
37                       11                    NaN         196   
58                       15                    NaN         196   
75                       11                    NaN         196   
...                     ...                    ...         ...   
32404835                 15                    NaN        9818   
32404840                 14                    NaN       15408   
32404843                 13                    NaN       43553   
32404856                  7                    NaN       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered  ... price_range_loc     busiest_day  \
0                         1          0  ...       Mid-range  Regularly busy   
23                        4          0  ...       Mid-range  Regularly busy   
37                        1          0  ...       Mid-range  Regularly busy   
58                        6          0  ...       Mid-range  Regularly busy   
75                        2          0  ...       Mid-range  Regularly busy   
...                     ...        ...  ...             ...             ...   
32404835                  9          0  ...       Mid-range  Regularly busy   
32404840                  6          0  ...       Mid-range  Regularly busy   
32404843                  2          0  ...       Low-range     Busiest day   
32404856                  4          0  ...       Mid-range  Regularly busy   
32404858                  8          0  ...       Mid-range      Least busy   

          busiest_days  busiest_period_of_day max_order      loyalty_flag  \
0         Regular days            Most orders        10      New customer   
23        Busiest days            Most orders        27  Regular customer   
37        Slowest days            Most orders        24  Regular customer   
58        Slowest days            Most orders        19  Regular customer   
75        Slowest days            Most orders         3      New customer   
...                ...                    ...       ...               ...   
32404835  Slowest days            Most orders        34  Regular customer   
32404840  Slowest days            Most orders        10      New customer   
32404843  Busiest days            Most orders        31  Regular customer   
32404856  Regular days         Average orders         3      New customer   
32404858  Slowest days            Most orders        16  Regular customer   

         average_spent spending_flag median_orders   order_frequency_flag  
0             6.367797   Low spender          20.5  Non-frequent customer  
23            6.356287   Low spender           8.0      Frequent customer  
37            7.656790   Low spender           7.0      Frequent customer  
58            6.441221   Low spender          11.5       Regular customer  
75            9.385714   Low spender          19.0       Regular customer  
...       

In [52]:
df_nan = ords_prods_merge[ords_prods_merge['order_id'] == 895835]

In [55]:
df_nan.head(20)

order_id  user_id eval_set  order_number  order_day_of_week  \
0     895835        1    prior             1                  2   
1     895835        1    prior             2                  3   
2     895835        1    prior             3                  3   
3     895835        1    prior             4                  4   
4     895835        1    prior             5                  4   
5     895835        1    prior             6                  2   
6     895835        1    prior             7                  1   
7     895835        1    prior             8                  1   
8     895835        1    prior             9                  1   
9     895835        1    prior            10                  4   
10    895835       15    prior            15                  1   
11    895835       15    prior            17                  2   
12    895835       15    prior            18                  2   
13    895835       15    prior            21                  1   
14    895835       15    prior            22                  1   
15    895835       19    prior             2                  5   
16    895835       19    prior             5                  6   
17    895835       19    prior             7                  4   
18    895835       21    prior            10                  3   
19    895835       31    prior            10                  3   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN         196                  1   
1                   7                   15.0         196                  1   
2                  12                   21.0         196                  1   
3                   7                   29.0         196                  1   
4                  15                   28.0         196                  1   
5                   7                   19.0         196                  1   
6                   9                   20.0         196                  1   
7                  14                   14.0         196                  2   
8                  16                    0.0         196                  4   
9                   8                   30.0         196                  1   
10                  9                    7.0         196                  2   
11                 16                    8.0         196                  6   
12                 11                    7.0         196                  1   
13                  9                    7.0         196                  1   
14                 10                   14.0         196                  1   
15                 14                    6.0         196                  3   
16                 12                    8.0         196                 15   
17                 17                    6.0         196                  1   
18                 10                    6.0         196                  2   
19                  8                    7.0         196                 17   

    reordered  ... price_range_loc     busiest_day  busiest_days  \
0           0  ...       Mid-range  Regularly busy  Regular days   
1           1  ...       Mid-range  Regularly busy  Slowest days   
2           1  ...       Mid-range  Regularly busy  Slowest days   
3           1  ...       Mid-range      Least busy  Slowest days   
4           1  ...       Mid-range      Least busy  Slowest days   
5           1  ...       Mid-range  Regularly busy  Regular days   
6           1  ...       Mid-range  Regularly busy  Busiest days   
7           1  ...       Mid-range  Regularly busy  Busiest days   
8           1  ...       Mid-range  Regularly busy  Busiest days   
9           1  ...       Mid-range      Least busy  Slowest days   
10          0  ...       Mid-range  Regularly busy  Busiest days   
11          1  ...       Mid-range  Regularly busy  Regular days   
12          1  ...       Mid-range  Regularly busy  

## Step 8 - Ensure your notebook is clean and structured and that your code is well commented.

## Step 9 - Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [63]:
# exporting dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_updated2.pkl'))